File: lib/kanban/store.ts
import { promises as fs } from "node:fs";
import path from "node:path";

import { DEFAULT_ENGINEERING_COLUMN_NAMES } from "@/lib/kanban/constants";
import type {
  KanbanBoard,
  KanbanColumn,
  KanbanWorkItem,
  UpdateColumnsInput,
} from "@/lib/kanban/types";
import {
  normalizeBoardName,
  normalizeColumnsInput,
  normalizeCreateWorkItemInput,
  normalizeUpdateWorkItemInput,
  validateStoredBoard,
} from "@/lib/kanban/validation";

const DATA_DIRECTORY = path.join(process.cwd(), "data");
const BOARD_FILE_PATH = path.join(
  DATA_DIRECTORY,
  "engineering-kanban-board.json",
);

export class KanbanStoreError extends Error {
  status: number;

  constructor(message: string, status = 400) {
    super(message);
    this.name = "KanbanStoreError";
    this.status = status;
  }
}

function nowIsoString() {
  return new Date().toISOString();
}

function sortBoard(board: KanbanBoard): KanbanBoard {
  return {
    ...board,
    columns: [...board.columns].sort((left, right) => left.order - right.order),
    items: [...board.items].sort((left, right) => left.order - right.order),
  };
}

async function ensureDataDirectory() {
  await fs.mkdir(DATA_DIRECTORY, { recursive: true });
}

async function writeBoard(board: KanbanBoard) {
  await ensureDataDirectory();
  const normalized = sortBoard(board);
  await fs.writeFile(BOARD_FILE_PATH, JSON.stringify(normalized, null, 2));
  return normalized;
}

function defaultColumns(): KanbanColumn[] {
  return DEFAULT_ENGINEERING_COLUMN_NAMES.map((name, index) => ({
    id: crypto.randomUUID(),
    name,
    order: index,
  }));
}

function nextItemOrder(items: KanbanWorkItem[], columnId: string) {
  const columnItems = items.filter((item) => item.columnId === columnId);

  if (columnItems.length === 0) {
    return 0;
  }

  return Math.max(...columnItems.map((item) => item.order)) + 1;
}

export async function loadBoard() {
  try {
    const fileContents = await fs.readFile(BOARD_FILE_PATH, "utf8");
    const parsed = JSON.parse(fileContents) as unknown;
    return sortBoard(validateStoredBoard(parsed));
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      return null;
    }

    throw new KanbanStoreError("Failed to load persisted board.", 500);
  }
}

export async function createBoard(nameInput: unknown) {
  const existing = await loadBoard();

  if (existing) {
    throw new KanbanStoreError("A board already exists.", 409);
  }

  const now = nowIsoString();
  const board: KanbanBoard = {
    id: crypto.randomUUID(),
    name: normalizeBoardName(nameInput),
    createdAt: now,
    updatedAt: now,
    columns: defaultColumns(),
    items: [],
  };

  return writeBoard(board);
}

async function requireBoard() {
  const board = await loadBoard();

  if (!board) {
    throw new KanbanStoreError("Create a board before updating it.", 404);
  }

  return board;
}

export async function saveColumns(input: UpdateColumnsInput["columns"]) {
  const board = await requireBoard();
  const normalizedColumns = normalizeColumnsInput(input);
  const persistedColumns = normalizedColumns.map((column, index) => ({
    id: column.id ?? crypto.randomUUID(),
    name: column.name,
    order: index,
  }));
  const nextColumnIds = new Set(persistedColumns.map((column) => column.id));
  const orphanedItems = board.items.filter(
    (item) => !nextColumnIds.has(item.columnId),
  );

  if (orphanedItems.length > 0) {
    throw new KanbanStoreError(
      "Move all work items out of a column before removing it.",
      409,
    );
  }

  return writeBoard({
    ...board,
    columns: persistedColumns,
    updatedAt: nowIsoString(),
  });
}

export async function createWorkItem(input: unknown) {
  const board = await requireBoard();
  const normalized = normalizeCreateWorkItemInput(input, board);
  const now = nowIsoString();

  const item: KanbanWorkItem = {
    id: crypto.randomUUID(),
    title: normalized.title,
    description: normalized.description ?? "",
    assignee: normalized.assignee ?? "",
    priority: normalized.priority ?? "medium",
    labels: normalized.labels ?? [],
    estimate: normalized.estimate ?? null,
    startDate: normalized.startDate ?? null,
    dueDate: normalized.dueDate ?? null,
    isBlocked: normalized.isBlocked ?? false,
    coverImage: normalized.coverImage ?? null,
    columnId: normalized.columnId,
    order: nextItemOrder(board.items, normalized.columnId),
    createdAt: now,
    updatedAt: now,
  };

  return writeBoard({
    ...board,
    items: [...board.items, item],
    updatedAt: now,
  });
}

export async function updateWorkItem(input: unknown) {
  const board = await requireBoard();
  const normalized = normalizeUpdateWorkItemInput(input, board);
  const current = board.items.find((item) => item.id === normalized.id);

  if (!current) {
    throw new KanbanStoreError("Work item not found.", 404);
  }

  const nextColumnId = normalized.columnId ?? current.columnId;
  const movedColumns = nextColumnId !== current.columnId;
  const nextOrder = movedColumns
    ? nextItemOrder(
        board.items.filter((item) => item.id !== current.id),
        nextColumnId,
      )
    : current.order;
  const updatedItem: KanbanWorkItem = {
    ...current,
    ...normalized,
    columnId: nextColumnId,
    order: nextOrder,
    updatedAt: nowIsoString(),
  };

  return writeBoard({
    ...board,
    items: board.items.map((item) =>
      item.id === current.id ? updatedItem : item,
    ),
    updatedAt: nowIsoString(),
  });
}

File: apps/api/store.py
from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Optional

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from models import KanbanBoard, KanbanColumn, KanbanWorkItem
from orm import BoardRow, ColumnRow, WorkItemRow
from validation import (
    normalize_board_name,
    normalize_columns_input,
    normalize_create_work_item_input,
    normalize_update_work_item_input,
)

DEFAULT_COLUMN_NAMES = [
    "Backlog",
    "Ready",
    "In Progress",
    "In Review",
    "Blocked",
    "Testing",
    "Done",
]


class KanbanStoreError(Exception):
    def __init__(self, message: str, status: int = 400):
        super().__init__(message)
        self.status = status


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _row_to_column(row: ColumnRow) -> KanbanColumn:
    return KanbanColumn(id=row.id, name=row.name, order=row.order)


def _row_to_item(row: WorkItemRow) -> KanbanWorkItem:
    return KanbanWorkItem(
        id=row.id,
        columnId=row.column_id,
        title=row.title,
        description=row.description,
        assignee=row.assignee,
        priority=row.priority,
        labels=row.labels or [],
        estimate=row.estimate,
        startDate=row.start_date,
        dueDate=row.due_date,
        isBlocked=row.is_blocked,
        coverImage=row.cover_image,
        order=row.order,
        createdAt=row.created_at,
        updatedAt=row.updated_at,
    )


async def _assemble_board(board_row: BoardRow, db: AsyncSession) -> KanbanBoard:
    cols_result = await db.execute(
        select(ColumnRow)
        .where(ColumnRow.board_id == board_row.id)
        .order_by(ColumnRow.order)
    )
    items_result = await db.execute(
        select(WorkItemRow)
        .where(WorkItemRow.board_id == board_row.id)
        .order_by(WorkItemRow.order)
    )
    return KanbanBoard(
        id=board_row.id,
        name=board_row.name,
        createdAt=board_row.created_at,
        updatedAt=board_row.updated_at,
        columns=[_row_to_column(r) for r in cols_result.scalars().all()],
        items=[_row_to_item(r) for r in items_result.scalars().all()],
    )


async def load_board(db: AsyncSession) -> Optional[KanbanBoard]:
    result = await db.execute(select(BoardRow).limit(1))
    board_row = result.scalars().first()
    if board_row is None:
        return None
    return await _assemble_board(board_row, db)


async def _require_board(db: AsyncSession) -> KanbanBoard:
    board = await load_board(db)
    if board is None:
        raise KanbanStoreError("Create a board before updating it.", 404)
    return board


async def create_board(name_input: object, db: AsyncSession) -> KanbanBoard:
    existing = await load_board(db)
    if existing:
        raise KanbanStoreError("A board already exists.", 409)

    now = _now()
    board_id = str(uuid.uuid4())
    db.add(BoardRow(id=board_id, name=normalize_board_name(name_input), created_at=now, updated_at=now))

    for index, col_name in enumerate(DEFAULT_COLUMN_NAMES):
        db.add(ColumnRow(id=str(uuid.uuid4()), board_id=board_id, name=col_name, order=index))

    await db.flush()

    result = await db.execute(select(BoardRow).where(BoardRow.id == board_id))
    return await _assemble_board(result.scalars().one(), db)


async def save_columns(columns_input: object, db: AsyncSession) -> KanbanBoard:
    board = await _require_board(db)

    if isinstance(columns_input, list):
        raw = [c.model_dump() if hasattr(c, "model_dump") else c for c in columns_input]
    else:
        raw = columns_input

    normalized = normalize_columns_input(raw)
    next_column_ids = {
        col["id"] if col["id"] else str(uuid.uuid4())
        for col in normalized
    }

    orphaned = [item for item in board.items if item.columnId not in next_column_ids]
    if orphaned:
        raise KanbanStoreError("Move all work items out of a column before removing it.", 409)

    await db.execute(delete(ColumnRow).where(ColumnRow.board_id == board.id))

    for index, col in enumerate(normalized):
        col_id = col["id"] if col["id"] else str(uuid.uuid4())
        db.add(ColumnRow(id=col_id, board_id=board.id, name=col["name"], order=index))

    now = _now()
    board_result = await db.execute(select(BoardRow).where(BoardRow.id == board.id))
    board_row = board_result.scalars().one()
    board_row.updated_at = now

    await db.flush()
    return await _assemble_board(board_row, db)


async def create_work_item(payload: object, db: AsyncSession) -> KanbanBoard:
    board = await _require_board(db)

    raw = payload.model_dump() if hasattr(payload, "model_dump") else payload
    normalized = normalize_create_work_item_input(raw, board)
    now = _now()

    column_items = [i for i in board.items if i.columnId == normalized["columnId"]]
    next_order = (max(i.order for i in column_items) + 1) if column_items else 0

    db.add(WorkItemRow(
        id=str(uuid.uuid4()),
        board_id=board.id,
        column_id=normalized["columnId"],
        title=normalized["title"],
        description=normalized.get("description") or "",
        assignee=normalized.get("assignee") or "",
        priority=normalized.get("priority") or "medium",
        labels=normalized.get("labels") or [],
        estimate=normalized.get("estimate"),
        start_date=normalized.get("startDate"),
        due_date=normalized.get("dueDate"),
        is_blocked=normalized.get("isBlocked") or False,
        cover_image=normalized.get("coverImage"),
        order=next_order,
        created_at=now,
        updated_at=now,
    ))

    board_result = await db.execute(select(BoardRow).where(BoardRow.id == board.id))
    board_row = board_result.scalars().one()
    board_row.updated_at = now

    await db.flush()
    return await _assemble_board(board_row, db)


async def delete_work_item(item_id: str, db: AsyncSession) -> KanbanBoard:
    board = await _require_board(db)

    item_result = await db.execute(
        select(WorkItemRow).where(WorkItemRow.id == item_id)
    )
    item_row = item_result.scalars().first()
    if item_row is None:
        raise KanbanStoreError("Work item not found.", 404)

    await db.execute(delete(WorkItemRow).where(WorkItemRow.id == item_id))

    now = _now()
    board_result = await db.execute(select(BoardRow).where(BoardRow.id == board.id))
    board_row = board_result.scalars().one()
    board_row.updated_at = now

    await db.flush()
    return await _assemble_board(board_row, db)


async def update_work_item(payload: object, db: AsyncSession) -> KanbanBoard:
    board = await _require_board(db)

    raw = payload.model_dump() if hasattr(payload, "model_dump") else payload
    normalized = normalize_update_work_item_input(raw, board)

    item_result = await db.execute(
        select(WorkItemRow).where(WorkItemRow.id == normalized["id"])
    )
    item_row = item_result.scalars().first()
    if item_row is None:
        raise KanbanStoreError("Work item not found.", 404)

    next_column_id = normalized.get("columnId", item_row.column_id)
    moved = next_column_id != item_row.column_id
    if moved:
        other_items = [i for i in board.items if i.id != item_row.id and i.columnId == next_column_id]
        next_order = (max(i.order for i in other_items) + 1) if other_items else 0
    else:
        next_order = item_row.order

    for field, col in [
        ("title", "title"),
        ("description", "description"),
        ("assignee", "assignee"),
        ("priority", "priority"),
        ("labels", "labels"),
        ("estimate", "estimate"),
        ("startDate", "start_date"),
        ("dueDate", "due_date"),
        ("isBlocked", "is_blocked"),
        ("coverImage", "cover_image"),
    ]:
        if field in normalized:
            setattr(item_row, col, normalized[field])

    item_row.column_id = next_column_id
    item_row.order = next_order
    item_row.updated_at = _now()

    board_result = await db.execute(select(BoardRow).where(BoardRow.id == board.id))
    board_row = board_result.scalars().one()
    board_row.updated_at = _now()

    await db.flush()
    return await _assemble_board(board_row, db)
